gadopt/nullspaces.py

```diff
@@ -0,0 +1,212 @@
+      translations: List of translations to include
+      ala_approximation: AnelasticLiquidApproximation for calculating (non-constant) right nullspace
+      top_subdomain_id: Boundary id of top surface. Required when providing
+                        ala_approximation.
+
+    Returns:
+      A Firedrake mixed vector space basis incorporating the null space components
+
+    """
+    # ala_approximation and top_subdomain_id are both needed when calculating right nullspace for ala
+    if (ala_approximation is None) != (top_subdomain_id is None):
+        raise ValueError(
+            "Both ala_approximation and top_subdomain_id must be provided, or both must be None."
+        )
+
+    stokes_subspaces = Z.subspaces
+
+    V_nullspace = create_u_nullspace(
+        stokes_subspaces[0],
+        rotational=rotational,
+        translations=translations)
+
+    if closed:
+        if ala_approximation:
+            p = ala_right_nullspace(
+                W=stokes_subspaces[1],
+                approximation=ala_approximation,
+                top_subdomain_id=top_subdomain_id,
+            )
+            p_nullspace = fd.VectorSpaceBasis([p], comm=Z.mesh().comm)
+            p_nullspace.orthonormalize()
+        else:
+            p_nullspace = fd.VectorSpaceBasis(constant=True, comm=Z.mesh().comm)
+    else:
+        p_nullspace = stokes_subspaces[1]
+
+    null_space = [V_nullspace, p_nullspace]
+
+    # If free surface unknowns, add dummy free surface nullspace
+    null_space += stokes_subspaces[2:]
+
+    return fd.MixedVectorSpaceBasis(Z, null_space)
+
+
+def create_u_nullspace(
+    V: fd.functionspaceimpl.WithGeometry,
+    rotational: bool = False,
+    translations: list[int] | None = None,
+    ala_approximation: AnelasticLiquidApproximation | None = None,
+    top_subdomain_id: str | int | None = None,
+) -> fd.nullspace.VectorSpaceBasis:
+    """Create a null space for the velocity (or displacement) in a Stokes system
+
+    Arguments:
+      V: Firedrake function space associated with the velocity or displacement
+      rotational: Whether to include all rotational modes
+      translations: List of translations to include
+
+    Returns:
+      A Firedrake vector space basis incorporating the null space components
+
+    """
+    X = fd.SpatialCoordinate(V.mesh())
+    dim = len(X)
+
+    if rotational:
+        if dim == 2:
+            rotV = fd.Function(V).interpolate(
+                fd.as_vector((-X[1], X[0]))
+            )
+            basis = [rotV]
+        elif dim == 3:
+            x_rotV = fd.Function(V).interpolate(
+                fd.as_vector((0, -X[2], X[1]))
+            )
+            y_rotV = fd.Function(V).interpolate(
+                fd.as_vector((X[2], 0, -X[0]))
+            )
+            z_rotV = fd.Function(V).interpolate(
+                fd.as_vector((-X[1], X[0], 0))
+            )
+            basis = [x_rotV, y_rotV, z_rotV]
+        else:
+            raise ValueError("Unknown dimension")
```

Comment:
```suggestion
            raise ValueError("Can only handle 2 or 3 dimensional spaces")
```
A more human readable error message.